### ISSUE TITLE
Circular dependencies get defined multiple times

### ISSUE THREAD

**Issue by Lakitna** [Swatinem/rollup-plugin-dts#39]
First of all: Great plugin :)

I'm working on a project (public soon) where I have the following pattern:

```typescript
import Foo from './Foo';

export default class FooManager {
    constructor() {
        this.foos = [ new Foo() ];
        // ...
    }
}
```

```typescript
import FooManager from './FooManager';

export default class Foo {
    public manager: FooManager;

    constructor(manager: FooManager) {
        this.manager = manager;

        // ...
    }
}
```

This way I can use config and such from `FooManager` in the methods of `Foo`.

This also creates a circular dependency, which works fine. When compiling the typescript down to javascript with `rollup-plugin-typescript2` things work as expected. But when compiling the types the following happens:

```typescript
declare class FooManager {
    // ...
}
declare class Foo {
    // ...
}
declare class FooManager$1 {
    // ...
}

export default FooManager$1;
```

Despite the unexpected output, it does work as expected when using it. So no rush in fixing this.

This happens when I use `FooManager.ts` as input, but also when I use the freshly generated `FooManager.d.ts` as input.

My `rollup.conf.js` for the sake of completeness:

```javascript
import pkg from './package.json';
import tslint from "rollup-plugin-tslint";
import typescript from 'rollup-plugin-typescript2';
import dts from 'rollup-plugin-dts';
import cleanup from 'rollup-plugin-cleanup';
import del from 'rollup-plugin-delete';

const config = [
    {
        input: './src/FooManager.ts',
        output: [
            { file: pkg.main, format: 'cjs' },
            { file: pkg.module, format: 'es' },
        ],
        plugins: [
            // Delete contents of target folder
            del({
                targets: pkg.files,
            }),

            // Lint source files
            tslint({
                throwOnError: true,
            }),

            // Compile source (typescript) to javascript
            typescript({
                tsconfig: './tsconfig.json',
            }),

            // Remove things like comments and whitespace
            cleanup({
                extensions: ['.ts', '.js'],
            }),
        ],
    },

    {
        input: './src/FooManager.ts',
        // input: './dist/src/FooManager.d.ts',
        output: [
            { file: pkg.types, format: 'es' }
        ],
        plugins: [
            // Generate types (.d.ts)
            dts(),
        ],
    },
];

module.exports = config;
```

Edit: Forgot to include the warning that's logged when bundling

```
./src/FooManager.ts → dist/FooManager.d.ts...
(!) Circular dependency: src\Foo.ts -> src\FooManager.ts -> src\Foo.ts
created dist/FooManager.d.ts in 1.3s
```

**Comment by Swatinem** [comment]
Hm, I can’t reproduce this in #40.

Are you on windows? Since the rollup warning uses backslashes as separators. I wonder if there is a problem around file name mapping? That it does not correctly sanitize `/` and `\` and thinks those are two different files? Not quite sure.

Would you mind running the testsuite from #40 on your machine? Or create a minimal testcase that shows this problem?

**Comment by Lakitna** [comment]
I am indeed on Windows, I'll try running #40 when I get home later tonight.

**Comment by Lakitna** [comment]
Hmmm so all tests fail...

I think it's due to line endings. Windows uses `\r\n` while Unix uses `\n`. I've got a single test case working: `call-signature` after changing the assertion on line 78 to this

```js
  expect(code.replace(/\r\n/g, '\n')).toEqual(expectedCode.replace(/\r\n/g, '\n'));
```

When also updating the expectation on line 88 I'm left with 2 failures. Both chunk name differences, one in `multiple-entries` and the other in `using-ts-files`. The test you added in #40 passes.
The closest I got to everything passing is when normalizing all line breaks to `\r`, so you must be using MacOS :)

Perhaps line endings should be normalized while bunding? Perhaps using something like https://www.npmjs.com/package/eol. That might keep everything stable on all platforms independent of them using `\r`, `\r\n` or `\n`.

**Comment by Swatinem** [comment]
> so you must be using MacOS :)

Oh hell no, I’d rather die :-D

But very interesting. I think you can configure git to not do line ending conversion automatically.
And yes, `multiple-entries` probably hashes the filenames as well, so it might be expected that it has different output depending on the directory separator.

Hm guess I have one more reason now to look into azure pipelines…

Would you mind putting a console log after this here:
https://github.com/Swatinem/rollup-plugin-dts/blob/d969ecba69431c7cdea1902f1eb8478beeb4c4d5/src/index.ts#L148-L152
`console.log({ source, importer, resolvedModule })`

I just saw that both `rollup-plugin-typescript{,2}` normalize the `importer`:

https://github.com/rollup/rollup-plugin-typescript/blob/6c433a6c482e3f36f01c83c5be130387da699c9c/src/index.js#L68
https://github.com/ezolenko/rollup-plugin-typescript2/blob/f262b559b59cda6e866eb199a11f0e174ffa09b0/src/index.ts#L124

I guess it could not hurt to do that as well. Does that help by any chance?

**Comment by Lakitna** [comment]
The result was the following:

```
  console.log src/index.ts:2560
    { source: './common',
      importer:
       'C:\\Users\\lakit\\Github\\rollup-plugin-dts\\tests\\testcases\\using-ts-files\\main-b.tsx',
      resolvedModule:
       { resolvedFileName:
          'C:/Users/lakit/Github/rollup-plugin-dts/tests/testcases/using-ts-files/common.ts',
         originalPath: undefined,
         extension: '.ts',
         isExternalLibraryImport: false,
         packageId: undefined } }
```

After normalizing `importer` I got

```
  console.log src/index.ts:2573
    { source: './common',
      importer:
       'C:/Users/lakit/Github/rollup-plugin-dts/tests/testcases/using-ts-files/main-b.tsx',
      resolvedModule:
       { resolvedFileName:
          'C:/Users/lakit/Github/rollup-plugin-dts/tests/testcases/using-ts-files/common.ts',
         originalPath: undefined,
         extension: '.ts',
         isExternalLibraryImport: false,
         packageId: undefined } }
```

But the tests still fail with the exact same pattern. (same hunk ids)

The plot thickens!

**Comment by Swatinem** [comment]
nice! Do you see anything unusual when running it on your own code?

**Comment by Lakitna** [comment]
The only thing I've noticed so far is what I described in my original description. I haven't been able to run it against the changes though. I'll look into that tomorrow after work. 

But right now it's time to visit my pillow :)

**Comment by Swatinem** [comment]
Tried your project on linux, after an initial fix https://github.com/Lakitna/lawful/pull/1 it builds just fine.
Maybe the case sensitivity is the cause here because it treated the two cases as different files?

**Comment by Lakitna** [comment]
Yeah I don't know how that name change reverted... I've run it on my work (windows) machine and I get the same output described above. I made a Gist so you can see what I see: https://gist.github.com/Lakitna/8d1b8c13ec92454e207169f12b48c5ab

Note the repeat of the class declaration for `Lawbook`. The second one is named `Lawbook$1`

**Comment by Lakitna** [comment]
Hmm, I dove a bit deeper in your code and I think I found the issue. I haven't been able to find where to fix it though.

When I add the following log to `index.ts`
```js
    renderChunk(code, chunk) {
      console.log({chunk});
``` 

The output is as follows
```js 
{ chunk:
   { code: undefined,
     dynamicImports: [],
     exports: [ 'default' ],
     facadeModuleId: 'C:\\Users\\lakit\\Github\\lawful\\src\\lawbook.ts',
     fileName: 'lawful.d.ts',
     imports: [],
     isDynamicEntry: false,
     isEntry: true,
     map: undefined,
     modules:
      [Object: null prototype] {
        'C:/Users/lakit/Github/lawful/src/config/types.ts': [Object],
        'C:/Users/lakit/Github/lawful/src/config/manager.ts': [Object],
        'C:/Users/lakit/Github/lawful/src/lawbook.ts': [Object],
        'C:/Users/lakit/Github/lawful/src/law.ts': [Object],
        'C:\\Users\\lakit\\Github\\lawful\\src\\lawbook.ts': [Object] },
     name: [Getter] } }
```

Note that `lawbook.ts` is in there twice, once with `\\` and once with `/`. 

The issue only arises when one of the circular dependencies is also the entry point. When I route the entry point through a wrapper (basically an `import x` > `export x` file) things do work as intended.

The question then becomes where we can access the absolute entry point path to normalise it.

**Comment by Swatinem** [comment]
Ah nice! I thought this would be an issue like this.

I’m actually surprised that `rollup-plugin-typescript{,2}` is not affected by the same issue, as their `resolveId` hook is more or less the same…
Would you mind logging the their `chunk` output as well?

**Comment by Lakitna** [comment]
Hmm it looks like `rollup-plugin-typescript2` does not use `renderChunk`. Though I'm seeing the same pattern where all paths use `/` except for the path of the entry file which uses `\\`.

Does this bug live in Rollup itself?

**Comment by Swatinem** [comment]
So I ended up setting up some basic dev environment on my windows, with git configured to *not* do `\n -> \r\n` conversion, and the TS compiler configured to output `\n`, all the tests worked fine.

https://github.com/Swatinem/rollup-plugin-dts/blob/d25ee6a15373545a39541bb03b1d0f6bb1829962/tests/testcases/tsconfig.json#L13

I then made sure to convert the result of `nodeModuleNameResolver` to platform native paths, which also fixes this issue. I will publish a new version as soon as I am back on linux :-)

Thank you very much @Lakitna for the help investigating this.
